[Mr-Teal] iP

Mr-Teal opened the pull request:
### Duke
This Duke may help you with managing all your tasks. It has
- easily used commands
- interesting interactions (T-Rex🦖 image)
- instant reaction

It may help you with:
- [x] Managing tasks
- [x] Differentiating task types
- [x] Search tasks by keywords
- [ ] Remind you with upcoming tasks (coming soon)

It supports:
1. normal tasks
2. tasks with due dates
3. tasks with starting and ending times

The `main` method is shown below:
```java
public class Duke {
    public static void main(String[] args) {
        new Duke("tasklist.txt").run();
    }
}
```
>Program based on [CS2103/T individual Project](https://github.com/nus-cs2103-AY2223S2/ip)

Commit:
Add Gradle support

Mr-Teal committed:
Duke.java: Complete Level-1

Mr-Teal committed:
Duke.java: Complete Level-2

Mr-Teal committed:
Duke.java: Complete Level-3; Create Task.java

Mr-Teal committed:
Duke.java: Complete Level-4; Create Todo.java, Deadline.java, and Event.java

Mr-Teal committed:
Add test with I/O redirection technique

Mr-Teal committed:
Duke.java: Complete Level-5; Create multiple Exception classes

Mr-Teal committed:
Duke.java: Complete Level-6; Create DeleteException.java

Mr-Teal committed:
Duke.java: Complete Level-7, Mass form changes; Add Save.java and TaskList.java; Delete unnecessary Exception classes

Mr-Teal committed:
Duke.java: Complete Level-8, Mass form changes; Delete unnecessary Exception classes

Mr-Teal committed:
Merge branch 'branch-Level-8'

Mr-Teal committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

Mr-Teal committed:
Duke.java: Complete A-MoreOOP; Create Commands.java, Parser.java, Storage.java and Ui.java; Delete Save.java

Mr-Teal committed:
Complete A-Packages; Create folders for different packages

Mr-Teal committed:
Complete A-Gradle

Mr-Teal committed:
Complete A-JUnit; Add multiple Test classes

Mr-Teal committed:
Complete A-JavaDoc; Add JavaDoc for most classes and methods

Mr-Teal committed:
Complete A-CodingStandard; Fix the code to fit the coding standard

Mr-Teal committed:
Complete Level-9; Create Find.java

Mr-Teal committed:
Complete A-CheckStyle; Make some coding style changes

Mr-Teal committed:
Complete Level-10; Add GUI; Mass form changes

Mr-Teal committed:
Merge branch 'branch-Level-10'

UnwilledTangent commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,58 @@
+ * The class used to identify user commands.
+ */
+public class Parser {
+    private enum cmds {
+        BYE, LIST, MARK, UNMARK, TODO, DEADLINE, EVENT, DELETE, FIND;
+    }
```
Love your use of enum

UnwilledTangent commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,66 @@
+package duke;
+
+import java.io.*;
```
Ohh... You are not supposed to use wildcard. Try to import only the parts of java.io that is relevant to your project. As per Java Coding Standard. https://se-education.org/guides/conventions/java/intermediate.html#package-and-import-statements

UnwilledTangent commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    /**
+     * Save the task list in the local disk.
```
```suggestion
     * Saves the task list in the local disk.
```

UnwilledTangent commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    /**
+     * Load the task list stored in the local disk.
```
```suggestion
     * Loads the task list stored in the local disk.
```

UnwilledTangent commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,38 @@
+    static final String LINE = "____________________________________________________________";
+
+    /**
+     * To welcome users.
```
```suggestion
     * Welcomes users
```
Refer to https://se-education.org/guides/conventions/java/intermediate.html#comments on how to code Javadocs. Should start with "Welcomes", not "Welcome" :)

UnwilledTangent commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    /**
+     * To read user input.
```
```suggestion
     * Reads user input.
```

UnwilledTangent commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,38 @@
+     *
+     * @return the user's input
+     */
+    public String readCMD() {
```
```suggestion
    public String readCmd() {
```
Abbreviations and acronyms should not be uppercase when used as a part of a name

UnwilledTangent commented on src/main/java/duke/commands/Bye.java:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    /**
+     * Say goodbye to the user.
```
```suggestion
     * Says goodbye to the user.
```

UnwilledTangent commented on src/main/java/duke/commands/Bye.java:
```diff
@@ -0,0 +1,38 @@
+     */
+    public String operate(TaskList lst, Ui ui, Storage storage) {
+        storage.save(lst);
+        return "I see you soon amigo I think! Roarrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrr!";
```
kawaii ne~

UnwilledTangent commented on src/main/java/duke/commands/Bye.java:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    /**
+     * Override the isBye() method as this is a bye command.
```
```suggestion
     * Overrides the isBye() method as this is a bye command.
```

UnwilledTangent commented on src/main/java/duke/commands/Command.java:
```diff
@@ -0,0 +1,30 @@
+    public abstract String operate(TaskList lst, Ui ui, Storage storage);
+
+    /**
+     * To check whether the command is a bye command.
```
```suggestion
     * Checks whether the command is a bye command.
```

UnwilledTangent commented on src/main/java/duke/commands/CreateDeadline.java:
```diff
@@ -0,0 +1,39 @@
+            if (cmdLine.length()<=9) throw new DukeException("Roarrrrrrrrrrrrrrrrr! Do you really have this Deadline task or not?");
+            String task = cmdLine.substring(9);
+            int pos = task.indexOf("/by");
+            if (pos==-1) throw new DukeException("Roarrrrrrrrrrrrrrrrr! Do you have a deadline for this task or not?");
```
```suggestion
            if (pos==-1) throw new DukeException("Rawrrrrrrrrrrrrrrrrr! Do you have a deadline for this task or not?");
```
jkjk

UnwilledTangent commented on src/main/java/duke/commands/notACommand.java:
```diff
@@ -0,0 +1,27 @@
+    }
+
+    /**
+     * Tell the user that he or she typed an invalid command.
```
```suggestion
     * Tells the user that he or she typed an invalid command.
```

UnwilledTangent commented on src/main/java/duke/taskType/Deadline.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    /**
+     * Override the toString() method to show the information of the Deadline task.
```
```suggestion
     * Overrides the toString() method to show the information of the Deadline task.
```

UnwilledTangent commented on src/main/java/duke/taskType/Event.java:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    /**
+     * Override the toString() method to show the information of the Event task.
```
```suggestion
     * Overrides the toString() method to show the information of the Event task.
```

UnwilledTangent commented on src/main/java/duke/taskType/Event.java:
```diff
@@ -0,0 +1,32 @@
+     * @param time1 the start time of the task
+     * @param time2 the end time of the task
+     */
+    public Event(String cont, String time1, String time2){
```
hmm... you might want to rename to startTime and endTime for clarity?

UnwilledTangent commented on src/main/java/duke/taskType/Task.java:
```diff
@@ -0,0 +1,53 @@
+ */
+public class Task implements Serializable {
+    protected String cont;
+    protected boolean finished;
```
any reason to use protected instead of private? :)

UnwilledTangent commented on src/main/java/duke/taskType/Task.java:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    /**
+     * Return whether the task is marked finished.
```
```suggestion
     * Returns whether the task is marked finished.
```

UnwilledTangent commented on src/main/java/duke/taskType/Task.java:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    /**
+     * Mark the task as unfinished.
```
```suggestion
     * Marks the task as unfinished.
```

UnwilledTangent commented on src/main/java/duke/taskType/Task.java:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    /**
+     * Override the toString() method to show the information of the task.
```
```suggestion
     * Overrides the toString() method to show the information of the task.
```

UnwilledTangent commented on src/main/java/duke/taskType/TaskList.java:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    /**
+     * Get the nth task in the list.
```
```suggestion
     * Gets the nth task in the list.
```

UnwilledTangent commented on src/main/java/duke/taskType/TaskList.java:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    /**
+     * Add a Task object in the list.
```
```suggestion
     * Adds a Task object in the list.
```

UnwilledTangent commented on src/main/java/duke/taskType/TaskList.java:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    /**
+     * Return how many tasks in the list.
```
```suggestion
     * Returns how many tasks in the list.
```

UnwilledTangent commented on src/main/java/duke/taskType/TaskList.java:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    /**
+     * Remove the task from the list.
```
```suggestion
     * Removes the task from the list.
```

UnwilledTangent commented on src/main/java/duke/taskType/TaskList.java:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    /**
+     * Override the toString() method to show the task list.
```
```suggestion
     * Overrides the toString() method to show the task list.
```

UnwilledTangent commented on src/main/java/duke/taskType/Todo.java:
```diff
@@ -0,0 +1,25 @@
+    }
+
+    /**
+     * Override the toString() method to show the information of the Todo task.
```
```suggestion
     * Overrides the toString() method to show the information of the Todo task.
```

UnwilledTangent commented on src/test/java/duke/commands/CreateTodoTest.java:
```diff
@@ -0,0 +1,20 @@
+import duke.*;
+import duke.commands.*;
+import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+public class CreateTodoTest {
```
```suggestion
import static org.junit.jupiter.api.Assertions.assertEquals;

public class CreateTodoTest {
```

UnwilledTangent commented on src/test/java/duke/commands/CreateTodoTest.java:
```diff
@@ -0,0 +1,20 @@
+package duke.commands;
+
+import duke.taskType.*;
+import duke.*;
+import duke.commands.*;
```
Don't use wildcards * bro/sis!

UnwilledTangent commented on src/test/java/duke/commands/UnmarkTest.java:
```diff
@@ -0,0 +1,23 @@
+package duke.commands;
+
+import duke.taskType.*;
+import duke.*;
+import duke.commands.*;
```
Try not to use wildcards *!

UnwilledTangent commented on src/test/java/duke/commands/UnmarkTest.java:
```diff
@@ -0,0 +1,23 @@
+import duke.*;
+import duke.commands.*;
+import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+public class UnmarkTest {
```
```suggestion
import static org.junit.jupiter.api.Assertions.assertEquals;

public class UnmarkTest {
```

UnwilledTangent commented on src/test/java/duke/taskType/DeadlineTest.java:
```diff
@@ -0,0 +1,14 @@
+import duke.*;
+import duke.commands.*;
+import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+public class DeadlineTest {
```
```suggestion
import static org.junit.jupiter.api.Assertions.assertEquals;

public class DeadlineTest {
```

UnwilledTangent commented on src/test/java/duke/taskType/EventTest.java:
```diff
@@ -0,0 +1,14 @@
+import duke.*;
+import duke.commands.*;
+import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+public class EventTest {
```
```suggestion
import static org.junit.jupiter.api.Assertions.assertEquals;

public class EventTest {
```

UnwilledTangent commented on src/test/java/duke/taskType/TodoTest.java:
```diff
@@ -0,0 +1,14 @@
+import duke.*;
+import duke.commands.*;
+import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+public class TodoTest {
```
```suggestion
import static org.junit.jupiter.api.Assertions.assertEquals;

public class TodoTest {
```

UnwilledTangent commented on src/test/java/duke/taskType/DeadlineTest.java:
```diff
@@ -0,0 +1,14 @@
+package duke.taskType;
+
+import duke.taskType.*;
+import duke.*;
+import duke.commands.*;
```
Wildcards *

UnwilledTangent commented on src/test/java/duke/taskType/EventTest.java:
```diff
@@ -0,0 +1,14 @@
+package duke.taskType;
+
+import duke.taskType.*;
+import duke.*;
+import duke.commands.*;
```
Wildcards *

UnwilledTangent commented on src/test/java/duke/taskType/TodoTest.java:
```diff
@@ -0,0 +1,14 @@
+package duke.taskType;
+
+import duke.taskType.*;
+import duke.*;
+import duke.commands.*;
```
Wildcards *

UnwilledTangent commented on text-ui-test/EXPECTED.TXT:
```diff
@@ -1,7 +1,38 @@
-| |_| | |_| |   <  __/
-|____/ \__,_|_|\_\___|
+____________________________________________________________
+Hello! I'm T-Rex. Roarrrrrrrrrrrrrr!
+What do you need from me?
+____________________________________________________________
 
+____________________________________________________________
+New Todo task is added. Roarrrrrrrrrrrrrrrrrrrrrrrrr!
+  [T][ ] borrow book
+You save 1 tasks in the list. Roarrrrrrrrrrrrrrrrrrrr!
+____________________________________________________________
+
+____________________________________________________________
+Roarrrrrrrrrrrrrrr! Task list shown below!
+1.[T][ ] borrow book
+____________________________________________________________
+
+____________________________________________________________
+New Deadline task is added. Roarrrrrrrrrrrrrrrrrrrrrrrrr!
+  [D][ ] return book (by: Sunday)
+You save 2 tasks in the list. Roarrrrrrrrrrrrrrrrrrrr!
+____________________________________________________________
+
+____________________________________________________________
+New Event task is added. Roarrrrrrrrrrrrrrrrrrrrrrrrr!
+  [E][ ] project meeting (from: Mon 2pm to: 4pm)
+You save 3 tasks in the list. Roarrrrrrrrrrrrrrrrrrrr!
+____________________________________________________________
+
+____________________________________________________________
+Roarrrrrrrrrrrrrrr! Task list shown below!
+1.[T][ ] borrow book
+2.[D][ ] return book (by: Sunday)
+3.[E][ ] project meeting (from: Mon 2pm to: 4pm)
+____________________________________________________________
+
+____________________________________________________________
+See you! Roarrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrr!
+____________________________________________________________
```
I think no need to add test files?

UnwilledTangent commented on text-ui-test/input.txt:
```diff
@@ -0,0 +1,6 @@
+todo borrow book
+list
+deadline return book /by Sunday
+event project meeting /from Mon 2pm /to 4pm
+list
+bye
```
I think no need to add test files? gitignore? :)

UnwilledTangent reviewed (commented):
I can see that you have made a very thorough effort for your iP! Keep up your good work! I just want to highlight to you that you should uses verbs like "Starts" or "Runs" for javadoc, and not "start" or "run", as per the Java Coding Standard. 

Also do remember to not use wildcards "*". "Rationale: Importing classes explicitly gives an excellent documentation value for the class at hand and makes the class easier to comprehend and maintain. Appropriate tools should be used in order to always keep the import list minimal and up to date. IDE's can be configured to do this easily."

Mr-Teal committed:
Complete A-Assertions: Add few assert()

Mr-Teal committed:
Complete A-CodeQuality
Improve the code quality based on the required version.
For Command class and its subclasses: use blank lines to make the operate() method more readable
For Find.java and Storage.java: Make the Happy Path prominent

Mr-Teal committed:
Merge pull request #1 from Mr-Teal/branch-A-Assertions

Complete A-Assertions: Add few assert()

Mr-Teal committed:
Merge pull request #2 from Mr-Teal/branch-A-CodeQuality

Complete A-CodeQuality

Mr-Teal committed:
Create github-actions-demo.yml

Mr-Teal committed:
Update github-actions-demo.yml

Delete the last three I/O redirection tests

Mr-Teal committed:
Merge pull request #3 from Mr-Teal/branch-A-CI

Create and update github-actions-demo.yml

Mr-Teal committed:
Update Storage.java

Delete a blank line which looks not nice

Mr-Teal committed:
Complete C-BetterSearch
(Acutally already implemented when completing Level-9 for Find command)
Mass code style change. Now no error in checkstyle.

Mr-Teal committed:
Merge pull request #4 from Mr-Teal/branch-BCD-Extension

Complete C-BetterSearch

Mr-Teal committed:
Rename notACommand.java to NotACommand.java

Mr-Teal committed:
Rename src/main/java/duke/taskType/Deadline.java to src/main/java/duke/tasktype/Deadline.java

Mr-Teal committed:
Rename src/main/java/duke/taskType/Event.java to src/main/java/duke/tasktype/Event.java

Mr-Teal committed:
Rename src/main/java/duke/taskType/Task.java to src/main/java/duke/tasktype/Task.java

Mr-Teal committed:
Rename src/main/java/duke/taskType/TaskList.java to src/main/java/duke/tasktype/TaskList.java

Mr-Teal committed:
Rename src/main/duke/taskType to src/main/duke/tasktype

Mr-Teal committed:
Rename src/test/java/duke/taskType/DeadlineTest.java to src/test/java/duke/tasktype/DeadlineTest.java

Mr-Teal committed:
Rename src/test/java/duke/taskType/EventTest.java to src/test/java/duke/tasktype/EventTest.java

Mr-Teal committed:
Rename src/test/java/duke/taskType/TodoTest.java to src/test/java/duke/tasktype/TodoTest.java

Mr-Teal committed:
Complete A-BetterGui
Add background picture
Improve the display of Duke

Mr-Teal committed:
Complete A-Personality
The T-Rex personality was added to my Duke long time ago.
This commit is used just to mention that the increment is finished.

Mr-Teal committed:
Update a screenshot of the UI of the product.

Mr-Teal committed:
Complete A-User

Mr-Teal committed:
Complete A-UserGuide

Update Readme.md in /docs

Mr-Teal committed:
Update README.md

Mr-Teal committed:
Update README.md

Mr-Teal committed:
Create the T-Rex.jar for user use.
